File: LeetCode/src/BinarySearchTree/KthSmallestElementInABST.java
package BinarySearchTree;

public class KthSmallestElementInABST {
	
	public static void main(String[] args) {
		
		Node root = null;
		//8, 3, 6, 10, 4, 7, 1, 14, 13
		root = CreateBinarySearchTree.insert(root, 8);
		root = CreateBinarySearchTree.insert(root, 3);
		root = CreateBinarySearchTree.insert(root, 6);
		root = CreateBinarySearchTree.insert(root, 1);
		root = CreateBinarySearchTree.insert(root, 4);
		root = CreateBinarySearchTree.insert(root, 7);
		root = CreateBinarySearchTree.insert(root, 10);
		root = CreateBinarySearchTree.insert(root, 14);
		root = CreateBinarySearchTree.insert(root, 13);
		
		Node kThSmallest = getkThSmallestElement(root, 7);
		if(kThSmallest != null) {
			System.out.println(kThSmallest.data);
			} else {
				System.out.println("The kth smallest element does not exist.");
			}
	}
	
	public static int countElement = 0;
	
	public static Node getkThSmallestElement(Node node, int k) {
		if(node == null) {
			return null;
		}
		
		Node left = getkThSmallestElement(node.left, k);
		
		if(left != null) {
			return left;
		}
		
		countElement++;
		
		if(countElement == k) {
			return node;
		}
		
		return getkThSmallestElement(node.right, k);
	}
}

File: Leetcode_Python/Neetcode150/KthSmallestElementInABST.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
        
class Solution:
    def kthSmallest(self, root: TreeNode, k: int) -> int:
        n = 0
        stack = []
        cur = root
        
        while True:
            while cur:
                stack.append(cur)
                cur = cur.left
                
            cur = stack.pop()
            n += 1
            
            if n == k:
                return cur.val
            
            cur = cur.right
            